### DISCUSSION TITLE
Chore: cleanup `TokensAnalyzer`

### DISCUSSION THREAD

**Discussion by SpacePossum** [PHP-CS-Fixer/PHP-CS-Fixer#7342] [RFC]
The `TokensAnalyzer` has, IMHO, to much responsibilities. Don't get me wrong, I really like it and all effort put into it is much appreciated by me :)
However, I would like to suggest some changes:
- move `getClassyElements` to `ClassyAnalyzer`
- move `getClassyModifiers` to `ClassyAnalyzer`
- move `isAnonymousClass` to `ClassyAnalyzer`
- move `isLambda` to `FunctionsAnalyzer`

**Comment by keradus** [comment]
I think TokensAnalyser become somehow god-object after we realised Tokens are too big already. We moved analysing to separated class, but again "one for all". Then we realised it's getting big and messy and started creating dedicated analyser-classes.

all in for cleanup (it could be public interface, so some deprecation may be needed first)

Yet, there is also concern of creating analyser class by different fixer again and again, which means those instances are acting independently. With that, there was suggestion to have `tokens` class create the analysers on-demand, but then have it cached for future usage.
Maybe worth to figure out this part, before moving analyser logic around?

**Reply by Wirone** [reply]
> Yet, there is also concern of creating analyser class by different fixer again and again, which means those instances are acting independently. With that, there was suggestion to have tokens class create the analysers on-demand, but then have it cached for future usage.

Instead of that, proper DI should be implemented, so any fixer could utilise any fine-crafted services (analyzers, modifiers etc.) in the same way.

**Comment by SpacePossum** [comment]
I get your take on this and agree.
Thinking forward loosely, maybe there is another way, having the analyzers injected into the fixers constructors. Downside would be that the analyzer methods would require the tokens collection as a param in (most of) their methods , yet these analyzers could still do some "caching" inside. Food for thought, now go and enjoy your holidays : D

**Reply by keradus** [reply]
here, we can find some details around the performance issue of not-cached analysers: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6939

- analysers injected to constructors (and tokens instance passed later) means that analyser has to take care of recognising the "same tokens instance/exact tokens -> reuse the cache", and putting cache burden to analysers
- moving instance creation to Tokens would solve passing Tokens instance to analysers nicely and cache is not a problem (we use same instance till tokens get modified), but would also create some coupling

> enjoy

too late.
